iframe content not painted when setting src on click event

### Summary

Dynamically setting an iframe (which is hidden with `visibility: collapse`)'s `src` in javascript, after which setting its visibility to `visibility: visible` makes the iframe's content not get painted instantly, but until it's invalidated.

The invisible iframe content is interactable though. (buttons can be clicked, text can be selected)

### Operating system

Linux

### Steps to reproduce

1. Open the linked repro URL in Ladybird
2. Press `Open iframe`
3. Look at the popup that holds the iframe


### Expected behavior

The green-lit "works" text should be visible instantly instead of it having to wait for the next repaint

### Actual behavior

The popup that holds the iframe was shown, but the iframe's contents were not rendered until I had resized my browser window.

### URL for a reduced test case

https://neoforean.github.io/repro/

### HTML/SVG/etc. source for a reduced test case

```html
N/A
```

### Log output and (if possible) backtrace


```shell
voi@pc:~/Documents/ladybird$ BUILD_PRESET=Debug ./Meta/ladybird.py run
ninja: Entering directory `/home/voi/Documents/ladybird/Build/debug'
[0/2] Re-checking globbed directories...
ninja: no work to do.
Fontconfig error: Cannot load default config file: No such file: (null)
Ladybird PID file '/run/user/1000/Ladybird.pid' exists with PID 297262, but process cannot be found
Fontconfig error: Cannot load default config file: No such file: (null)
Fontconfig error: Cannot load default config file: No such file: (null)
Fontconfig error: Cannot load default config file: No such file: (null)
Fontconfig error: Cannot load default config file: No such file: (null)
GrVkCommandPool: 94723509842112 (2 refs)
GrVkCommandPool: 139786086061248 (2 refs)
GrVkImageView: 139786086273232 (1 refs)
GrVkImage: 139786086125472 (1 refs)
GrVkImageView: 139786086342320 (1 refs)
GrVkImage: 139786086339568 (1 refs)
GrVkImageView: 139786086357840 (1 refs)
GrVkImageView: 139786086358528 (1 refs)
GrVkImage: 139786086355808 (1 refs)
GrVkRenderPass: 139786086069952 (4 refs)
GrVkFramebuffer: 139786086355584 (1 refs)
GrVkRenderPass: 139786086082272 (1 refs)
GrVkFramebuffer: 139786086388336 (1 refs)
GrVkPipeline: 139786086884608 (1 refs)
GrVkDescriptorPool: 139786086904928, type 1 (2 refs)
GrVkDescriptorSet: 139786086905112 (1 refs)
GrVkSampler: 139786086707600 (1 refs)
GrVkCommandPool: 139786086892416 (1 refs)
GrVkFramebuffer: 139786086871968 (1 refs)
../src/41119015ca-b0525dd526.clean/src/gpu/ganesh/GrManagedResource.h:61: fatal error: "check(0 == fHashSet.count())"

CRASH: Received signal SIGABRT (6)
Fontconfig error: Cannot load default config file: No such file: (null)
Fontconfig error: Cannot load default config file: No such file: (null)
Stack trace (most recent call first):
#0  0x00007f230ce2728f at /lib64/libc.so.6
#1  0x00007f230ce813cc in __pthread_kill_implementation at /lib64/libc.so.6
#2  0x00007f230ce2715d in gsignal at /lib64/libc.so.6
#3  0x00007f230ce0e6cf in abort at /lib64/libc.so.6
#4  0x00007f230b7c9bed in sk_abort_no_print() at /home/voi/Documents/ladybird/Build/vcpkg/buildtrees/skia/x64-linux-dynamic-dbg/../src/41119015ca-b0525dd526.clean/src/ports/SkMemory_malloc.cpp:62:10
#5  0x00007f230bc2a252 in GrManagedResource::Trace::~Trace()::{lambda()#1}::operator()() const at /home/voi/Documents/ladybird/Build/vcpkg/buildtrees/skia/x64-linux-dynamic-dbg/../src/41119015ca-b0525dd526.clean/src/gpu/ganesh/GrManagedResource.h:61:13
#6  0x00007f230bc2a296 in GrManagedResource::Trace::~Trace() at /home/voi/Documents/ladybird/Build/vcpkg/buildtrees/skia/x64-linux-dynamic-dbg/../src/41119015ca-b0525dd526.clean/src/gpu/ganesh/GrManagedResource.h:61:13
#7  0x00007f230ce299e0 in __run_exit_handlers at /lib64/libc.so.6
#8  0x00007f230ce29abd in exit at /lib64/libc.so.6
#9  0x00007f230ce105bb in __libc_start_call_main at /lib64/libc.so.6
#10 0x00007f230ce10667 in __libc_start_main_alias_2 at /lib64/libc.so.6
#11 0x0000562660f22ba4 in _start at /home/voi/Documents/ladybird/Build/debug/libexec/WebContent
19270.810 WebContent(298548): ImageDecoderClient: Failed to decode image with request token 0: Decoding failed: ICO size too large
```

### Screenshots or screen recordings

Firefox & Chromium-based:

https://github.com/user-attachments/assets/4482b27f-e55c-4b14-958e-c17bcd8349a0

Ladybird:

https://github.com/user-attachments/assets/8697e6a6-7208-4378-8256-bf423cc8a7e6

### Build flags or config settings

```BUILD_PRESET=Debug ./Meta/ladybird.py run```

### Contribute a patch?

- [ ] I’ll contribute a patch for this myself.

--- a/Libraries/LibWeb/DOM/Document.cpp
+++ b/Libraries/LibWeb/DOM/Document.cpp
@@ -7948,10 +7948,6 @@ void Document::set_needs_repaint(InvalidateDisplayList should_invalidate_display
 {
     auto navigable = this->navigable();
 
-    // OPTIMIZATION: Ignore set_needs_repaint() inside navigable containers (i.e frames) with visibility: hidden.
-    if (navigable && navigable->has_inclusive_ancestor_with_visibility_hidden())
-        return;
-
     if (should_invalidate_display_list == InvalidateDisplayList::Yes) {
         set_needs_to_record_display_list();
     }

--- a/Libraries/LibWeb/HTML/EventLoop/EventLoop.cpp
+++ b/Libraries/LibWeb/HTML/EventLoop/EventLoop.cpp
@@ -534,6 +534,10 @@ void EventLoop::update_the_rendering()
         auto navigable = doc->navigable();
         if (!navigable->needs_repaint())
             continue;
+        // OPTIMIZATION: Don't paint navigables hidden by an ancestor iframe with visibility: hidden.
+        //               needs_repaint() stays true — so, once the navigable becomes visible, it's painted.
+        if (navigable->has_inclusive_ancestor_with_visibility_hidden())
+            continue;
         if (navigable->is_svg_page())
             continue;
         if (auto document = navigable->active_document())

--- a/Tests/LibWeb/Ref/expected/iframe-repaint-after-visibility-hidden-ref.html
+++ b/Tests/LibWeb/Ref/expected/iframe-repaint-after-visibility-hidden-ref.html
@@ -1,0 +1,19 @@
+<!DOCTYPE html>
+<html>
+    <head>
+        <style>
+            body {
+                margin: 0;
+            }
+            iframe {
+                display: block;
+                width: 100px;
+                height: 100px;
+                border: 0;
+            }
+        </style>
+    </head>
+    <body>
+        <iframe srcdoc="<body style='background: green'></body>"></iframe>
+    </body>
+</html>

--- a/Tests/LibWeb/Ref/input/iframe-repaint-after-visibility-hidden.html
+++ b/Tests/LibWeb/Ref/input/iframe-repaint-after-visibility-hidden.html
@@ -1,0 +1,46 @@
+<!DOCTYPE html>
+<html class="reftest-wait">
+    <head>
+        <link rel="match" href="../expected/iframe-repaint-after-visibility-hidden-ref.html" />
+        <style>
+            body {
+                margin: 0;
+            }
+            iframe {
+                display: block;
+                width: 100px;
+                height: 100px;
+                border: 0;
+            }
+        </style>
+    </head>
+    <body>
+        <!-- Regression test for https://github.com/LadybirdBrowser/ladybird/issues/9305 -->
+        <!-- An iframe whose content changes while it has a visibility:hidden ancestor must be -->
+        <!-- painted once that ancestor becomes visible again. -->
+        <iframe id="frame" srcdoc="<body style='background: red'></body>"></iframe>
+        <script>
+            const frame = document.getElementById("frame");
+            let phase = 0;
+            frame.addEventListener("load", () => {
+                if (phase === 0) {
+                    // The red document has loaded and been painted; hide the iframe and swap in green.
+                    phase = 1;
+                    requestAnimationFrame(() => requestAnimationFrame(() => {
+                        frame.style.visibility = "hidden";
+                        frame.srcdoc = "<body style='background: green'></body>";
+                    }));
+                } else if (phase === 1) {
+                    // The green document loaded while the iframe was hidden; reveal it again.
+                    phase = 2;
+                    requestAnimationFrame(() => requestAnimationFrame(() => {
+                        frame.style.visibility = "visible";
+                        requestAnimationFrame(() => requestAnimationFrame(() => {
+                            document.documentElement.className = "";
+                        }));
+                    }));
+                }
+            });
+        </script>
+    </body>
+</html>

